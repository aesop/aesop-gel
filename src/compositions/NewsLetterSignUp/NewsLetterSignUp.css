@import 'styles/index';

.base {
  position: relative;
}

.signUpIcon,
.loading {
  fill: var(--color-error);
  &.dark {
    fill: var(--color-dark-copy);
  }

  &.light {
    fill: var(--color-light-copy);
  }
}

.inputField {
  position: relative;
  display: flex;
  width: 100%;
  align-items: flex-end;
}

.footerNewsletterInputWrapper {
  flex: 1 0 auto;
}

.button {
  height: 37px;
  box-sizing: border-box;
  flex: 0 0 auto;
  border: 1px solid var(--color-dark-copy);
  border-left: 0;
  margin-left: 0;
  background: transparent;
  border-radius: 0 3px 3px 0;
  cursor: pointer;
  font-size: rem(20px);

  &.dark {
    border-color: var(--color-dark-copy);
  }

  &.light {
    border-color: var(--color-light-copy);
  }

  &.hasError {
    border-color: var(--color-error);
  }

  &:focus {
    outline: none;
  }
}

.footerNewsLetter {
  height: 37px;
  border-right: 0;
  border-radius: 3px 0 0 3px;
}

.loading {
  position: absolute;
  top: 18px;
}

.subscriptionMessage {
  margin-top: 7px;
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(14px);
  line-height: 2.14;
  &.dark {
    color: var(--color-dark-copy);
  }

  &.light {
    color: var(--color-light-copy);
  }
}

.footerNewsletterErrorMessage {
  color: var(--color-error);
}
