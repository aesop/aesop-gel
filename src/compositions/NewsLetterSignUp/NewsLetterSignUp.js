import React, { useState } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import Icon from '~/components/Icon';
import Loading from '~/components/Loading';
import Checkbox from '~/components/Checkbox';
import TextInput from '~/components/TextInput';
import styles from './NewsLetterSignUp.css';
import { useNotificationContext } from 'contexts';

const NewsLetterSignUp = ({
  className,
  consentErrorMsg,
  errorMessage,
  onClick,
  showTermsConditionsTextBox,
  termsAndCondition,
  termsMessage,
  theme,
  subscriptionMessage,
}) => {
  const [hasContent, setHasContent] = useState(false);
  const [email, setEmail] = useState(undefined);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setErrorMessage] = useState(undefined);
  const [hasAcceptedSubscription, setHasAcceptedSubscription] = useState(false);
  const [hasAcceptedTerms, setHasAcceptedTerms] = useState(false);

  const inputFieldClassSet = cx(styles.inputField, styles[theme], className);
  const buttonClassSet = cx(
    styles.button,
    styles[theme],
    !!error && styles.hasError,
  );
  const signUpIconClassSet = cx(styles.signUpIcon, styles[theme]);
  const loadingClassSet = cx(styles.loading, styles[theme]);
  const subscriptionMessageClassSet = cx(
    styles.subscriptionMessage,
    styles[theme],
  );

  const footerSuccessModal = useNotificationContext();
  const { actionType, dispatch } = footerSuccessModal;

  const validate = value => {
    const EMAIL_REGEXP = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i;
    const string = String(value);
    if (!string) {
      return true;
    }

    const isValid = !!string.match(EMAIL_REGEXP);
    return isValid;
  };

  const onChange = e => {
    setEmail(e.target.value);
    setHasContent(e.target.value && e.target.value.trim() !== '');
  };

  const handleSubmit = async e => {
    e.preventDefault();
    setErrorMessage(undefined);
    const isValid = validate(email);
    if (
      isValid &&
      !(
        showTermsConditionsTextBox &&
        (!hasAcceptedSubscription || !hasAcceptedTerms)
      )
    ) {
      setIsLoading(true);
      await onClick(email);
      setIsLoading(false);
      dispatch({ type: actionType });
    } else {
      let errorMsg = errorMessage;
      if (
        showTermsConditionsTextBox &&
        (!hasAcceptedSubscription || !hasAcceptedTerms)
      ) {
        errorMsg = consentErrorMsg;
      }
      setErrorMessage(errorMsg);
    }
  };

  return (
    <div className={styles.base}>
      <form onSubmit={handleSubmit}>
        <div className={inputFieldClassSet}>
          <div className={styles.footerNewsletterInputWrapper}>
            <TextInput
              hasContent={hasContent}
              hasError={!!error}
              id="newsletter-email"
              inputClassName={cx(styles.footerNewsLetter)}
              isEnabled={!isLoading}
              label="Email Address"
              name="newsletter-email"
              onChange={onChange}
              theme={theme}
            />
          </div>
          <button
            className={buttonClassSet}
            disabled={!!isLoading || !hasContent}
          >
            <Icon
              className={signUpIconClassSet}
              name="rightArrow"
              theme={theme}
            />
          </button>
          {!!isLoading && (
            <Loading
              className={loadingClassSet}
              isLoading={true}
              theme={theme}
            />
          )}
        </div>
        <div className={styles.footerNewsletterErrorMessage}>{error}</div>
        {showTermsConditionsTextBox ? (
          <section>
            <Checkbox
              content={subscriptionMessage}
              id="subscription"
              isEnabled={!isLoading}
              onChange={() =>
                setHasAcceptedSubscription(!hasAcceptedSubscription)
              }
              theme={theme}
            />
            <Checkbox
              content={termsMessage}
              id="terms"
              isEnabled={!isLoading}
              onChange={() => setHasAcceptedTerms(!hasAcceptedTerms)}
              theme={theme}
            />
            <div className={styles.termsAndConditionWrapper}>
              <div className={styles.termsAndCondition}>
                {termsAndCondition}
              </div>
            </div>
          </section>
        ) : (
          <div className={subscriptionMessageClassSet}>
            {subscriptionMessage}
          </div>
        )}
      </form>
    </div>
  );
};

NewsLetterSignUp.propTypes = {
  className: PropTypes.string,
  consentErrorMsg: PropTypes.string,
  errorMessage: PropTypes.string,
  onClick: PropTypes.func,
  showTermsConditionsTextBox: PropTypes.bool,
  termsAndCondition: PropTypes.string,
  termsMessage: PropTypes.string,
  theme: PropTypes.string,
  subscriptionMessage: PropTypes.string,
};

NewsLetterSignUp.defaultProps = {
  className: undefined,
  consentErrorMsg: undefined,
  errorMessage: 'NEWSLETTER_SIGNUP',
  onClick: undefined,
  showTermsConditionsTextBox: false,
  termsAndCondition: undefined,
  termsMessage: undefined,
  theme: 'dark',
  subscriptionMessage: undefined,
};

export default NewsLetterSignUp;
