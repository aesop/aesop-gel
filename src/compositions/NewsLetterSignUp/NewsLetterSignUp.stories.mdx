import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import Podium from '~/components/Podium';
import NotificationModal from '~/components/NotificationModal';
import NotificationModalFixture from '~/components/NotificationModal/NotificationModal.fixture';
import NewsLetterSignUp from './NewsLetterSignUp';
import NewsLetterSignUpFixture from './NewsLetterSignUp.fixture'
import newsLetterSignUpOnClick from './__mocks__/NewsLetterSignUp.onClick';
import { NotificationContextProvider } from '~/contexts';

# NewsLetterSignUp

<Meta title="Compositions/NewsLetterSignUp" component={NewsLetterSignUp} />

<Canvas>
  <Story name="Base component">
    <Podium horizontalPadding="small" verticalPadding="small">
      <NotificationContextProvider>
        <div style={{position: 'relative'}}>
          <NotificationModal
            backgroundColor="#d5d5cc"
            notificationMessage={NotificationModalFixture.notificationMessage}
          />
          <NewsLetterSignUp
            consentErrorMsg={NewsLetterSignUpFixture.consentErrorMsg}
            errorMessage={NewsLetterSignUpFixture.errorMessage}
            onClick={newsLetterSignUpOnClick}
            showTermsConditionsTextBox={knobs.boolean('showTermsConditionsTextBox', true, 'Presentation')}
            subscriptionMessage={NewsLetterSignUpFixture.subscriptionMessage}
            termsAndCondition={NewsLetterSignUpFixture.termsAndCondition}
            termsMessage={NewsLetterSignUpFixture.termsMessage}
            theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
          />
        </div>
      </NotificationContextProvider>
    </Podium>
  </Story>
</Canvas>

## API

<ArgsTable of={NewsLetterSignUp} />

