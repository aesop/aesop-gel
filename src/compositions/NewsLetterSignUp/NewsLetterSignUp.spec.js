import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import NewsLetterSignUp from './NewsLetterSignUp';
import NewsLetterSignUpFixture from './NewsLetterSignUp.fixture';

configure({ adapter: new Adapter() });

describe('<NewsLetterSignUp />', () => {
  it('should be defined', () => {
    expect(NewsLetterSignUp).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <NewsLetterSignUp
          consentErrorMsg={NewsLetterSignUpFixture.consentErrorMsg}
          errorMessage={NewsLetterSignUpFixture.errorMessage}
          showTermsConditionsTextBox={true}
          subscriptionMessage={NewsLetterSignUpFixture.subscriptionMessage}
          termsAndCondition={NewsLetterSignUpFixture.termsAndCondition}
          termsMessage={NewsLetterSignUpFixture.termsMessage}
          theme="light"
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
