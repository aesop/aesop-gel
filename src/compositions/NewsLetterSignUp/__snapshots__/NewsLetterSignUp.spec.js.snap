// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsLetterSignUp /> renders base component correctly 1`] = `
<div
  className="base"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="inputField light"
    >
      <div
        className="footerNewsletterInputWrapper"
      >
        <div
          className="formText"
        >
          <label
            htmlFor="newsletter-email"
          >
            <input
              className="formTextInput light footerNewsLetter"
              disabled={false}
              id="newsletter-email"
              name="newsletter-email"
              noValidate={true}
              onChange={[Function]}
              placeholder=""
              type="text"
            />
            <span
              className="formTextLabel light"
            >
              Email Address
            </span>
          </label>
        </div>
      </div>
      <button
        className="button light"
        disabled={true}
      >
        <svg
          className="base rightArrow light signUpIcon light"
          data-ref=""
          focusable="false"
          height={12}
          role="img"
          style={
            Object {
              "height": "12px",
              "width": "12px",
            }
          }
          tabIndex={-1}
          viewBox="0 0 50 50"
          width={12}
        >
          <g>
            <path
              d="M30.1,5.3L50,25.1L30.1,45h-6.6l18-17.6H0v-4.8h41.5l-18-17.6h6.6V5.3z"
            />
          </g>
        </svg>
      </button>
    </div>
    <div
      className="footerNewsletterErrorMessage"
    />
    <section>
      <label
        className="base"
        htmlFor="subscription"
      >
        <input
          className="checkBox light"
          disabled={false}
          id="subscription"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="content light"
        >
          Subscribe to communications about Aesop products, services, stores, events and matters of cultural interest.
        </span>
      </label>
      <label
        className="base"
        htmlFor="terms"
      >
        <input
          className="checkBox light"
          disabled={false}
          id="terms"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="content light"
        >
          I accept the Terms & Conditions 
        </span>
      </label>
      <label
        htmlFor="term-condition"
      >
        <div
          className="base light"
        >
          <textarea
            className="textarea light"
            disabled={true}
            id="term-condition"
            rows={3}
            value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta."
          />
        </div>
      </label>
    </section>
  </form>
</div>
`;
